File: projeto/frontend/site/vox/src/app/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Notification {
  id: number;
  message: string;
  read: boolean;
  createdAt: string;
}

@Injectable({ providedIn: 'root' })
export class NotificationService {
  private readonly API_URL = 'http://localhost:8080';

  constructor(private http: HttpClient) {}

  getNotifications(): Observable<Notification[]> {
    return this.http.get<Notification[]>(`${this.API_URL}/api/notifications`);
  }

  getUnreadNotifications(): Observable<Notification[]> {
    return this.http.get<Notification[]>(`${this.API_URL}/api/notifications/unread`);
  }

  getUnreadCount(): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.API_URL}/api/notifications/count`);
  }

  markAsRead(id: number): Observable<void> {
    return this.http.patch<void>(`${this.API_URL}/api/notifications/${id}/read`, null);
  }

  markAllAsRead(): Observable<void> {
    return this.http.patch<void>(`${this.API_URL}/api/notifications/read-all`, null);
  }
}

File: projeto/frontend/app/app_vox/lib/services/notification_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_client.dart';
import '../models/notification.dart';

class NotificationService {
  Future<List<AppNotification>> getNotifications() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/notifications'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => AppNotification.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<AppNotification>> getUnreadNotifications() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/notifications/unread'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final list = jsonDecode(response.body) as List;
    return list
        .map((e) => AppNotification.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<int> getUnreadCount() async {
    final response = await http.get(
      Uri.parse('${ApiClient.baseUrl}/api/notifications/count'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
    final data = jsonDecode(response.body) as Map<String, dynamic>;
    return data['count'] as int? ?? 0;
  }

  Future<void> markAsRead(int id) async {
    final response = await http.patch(
      Uri.parse('${ApiClient.baseUrl}/api/notifications/$id/read'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }

  Future<void> markAllAsRead() async {
    final response = await http.patch(
      Uri.parse('${ApiClient.baseUrl}/api/notifications/read-all'),
      headers: await ApiClient.authHeaders(),
    );
    ApiClient.checkResponse(response);
  }
}
